Feature bjs2 77154 education

FilippovichSergey opened the pull request.

FilippovichSergey committed:
created classes EducationDto, EducationService, EducationController, and DataValidationException (for exception)
created interface EducationMapper
updated class Education (added @Setter)

FilippovichSergey committed:
updated classes EducationService and EducationController and interface EducationMapper

FilippovichSergey committed:
interface EducationMapper

FilippovichSergey committed:
updated class EducationDto and interface EducationMapper

re1971reg commented on src/main/java/school/faang/user_service/controller/education/EducationController.java:
```diff
@@ -0,0 +1,30 @@
+    private final EducationService educationService;
+
+    @Autowired
+    public EducationController(EducationService educationService) {
```
В данном случае можно воспользоваться библиотекой lombok и указав аннотацию `@RequiredArgsConstructor` конструктор не придётся писать самому. Все переменные класса, которые объявлены как `private final` будут подставлены в конструктор автоматически.
посмотри здесь примеры на [lombok на habr.com](https://habr.com/ru/companies/piter/articles/676394/)

Так же, если у тебя один конструктор, то аннотацию `@Autowired` писать не обязательно.

re1971reg commented on src/main/java/school/faang/user_service/service/education/DataValidationException.java:
классы исключений лучше располагать в специальном каталоге для исключений, в нашем случае это каталог exception, который должен находиться на уровне каталогов entity, service, controller. Если влить к себе ветку [chimera-master-stream10](https://github.com/CorporationX/user_service/tree/chimera-master-stream10), то этот каталог появится сам. Влить ветку master лучше конечно через rebase. Рекоментации по rebase от Константина [здесь в discord](https://discord.com/channels/1102662981743030334/1126205623265140736/1372497620999868457). Что такое rebase и для чего он нужен показывается [здесь на faang-school.com](https://faang-school.com/courses/4jnzmndg/w8zx6xwz)

re1971reg commented on src/main/java/school/faang/user_service/entity/Education.java:
```diff
@@ -8,12 +8,10 @@
-import lombok.NoArgsConstructor;
+import lombok.*;
 
 @Getter
```
Аннотациии `@Setter` и `@Getter` замени на `@Data`
Объяснения с примерами [lombok на habr.com](https://habr.com/ru/companies/piter/articles/676394/) или здесь на [faang-school.com](https://faang-school.com/courses/g89z8bzp/w8zxmw3d)

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,74 @@
+    private final EducationMapper educationMapper;
+
+    @Autowired
+    public EducationService(UserRepository userRepository, EducationRepository educationRepository, EducationMapper educationMapper) {
```
комментарий такой же как и в классе EducationController.java. Лучше воспользоваться библиотекой lombok, кода писать меньше.

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public void checkYearFrom(EducationDto educationDto) throws DataValidationException {
+        if (educationDto.getYearFrom() > LocalDate.now().getYear()) {
```
Сравнение значений объектов надо делать через метод `compare`, а этот код работать не будет.

re1971reg commented on src/main/java/school/faang/user_service/controller/education/EducationController.java:
```diff
@@ -0,0 +1,30 @@
+        this.educationService = educationService;
+    }
+
+    public EducationDto addEducation(long userId, EducationDto educationDto) throws DataValidationException {
```
переменную `long userId` здесь и в других методах лучше изменить на `Long userId`. Если при вызове метода в переменной `userId` укажут null, то произойдёт автоматическое преобразование в 0.
А методе сервиса `EducationService.java` реализовать проверку на null.

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,74 @@
+        Education education = educationMapper.toEducation(educationDto);
+        education.setUser(userOptional.get());
+
+        educationRepository.save(education);
```
метод `save` возвращает переменную типа `Education`, которая обновляется значениями из БД после сохранения, в твоём случае появляется id, т.е. :
```
Education resultEducation = educationRepository.save(education);
return educationMapper.toEducationDto(resultEducation);
```

FilippovichSergey committed:
classes EducationController and EducationService: added the @RequiredArgsContructor annotation
class DataValidationException: moved to a new package "exception"
class Education: changed annotations @Getter and @Setter to @Data
class EducationController and EducationService: changed "long" to "Long" for ids.
class EducationService: added the method for checking if id is null.
class EducationService, method checkYearFrom: added "compare" method.
class EducationService, method saveEducation: updated the return data.

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,80 @@
+
+    public void checkUserIdNull(Long userId) throws DataValidationException {
+        Optional<User> userOptional = userRepository.findById(userId);
+        if (userOptional.isPresent()) {
```
здесь ты наверное хотел выкинуть ошибку при отсутсвии пользователя? а у тебя получилось, что ошибка будет при наличии пользователя.
так же посмотри комментарий к методу `checkUserIdEmpty`.

если же ты хотел проверить, что userId не null, то и надо было просто написать:
```
if (userId == null) {
    throw new DataValidationException("User ID is null.");
}
```

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,80 @@
+    }    //However, the yearFrom can be equal to the current year.
+
+    public Optional<User> checkUserIdEmpty(Long userId) throws DataValidationException {
+        Optional<User> userOptional = userRepository.findById(userId);
```
всю конструкцию получения пользователя и выбрасывания сообщения при его отсутсвии делается так же как у тебя написано в методе
`public EducationDto getById(Long educationId)`
т.е.
```
public User checkUserIdEmpty(Long userId) throws DataValidationException {
    User user = userRepository.findById(userId)
        .orElseThrow(() -> new DataValidationException("User with ID " + userId + " not found"));
    return user;
}
```

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public EducationDto addEducation(Long userId, EducationDto educationDto) throws DataValidationException {
+        checkYearFrom(educationDto);
```
этот метод есть в `saveEducation()` и поэтому его лучше отсюда убрать и оставить только в одном месте, в методе `saveEducation()`

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,80 @@
+
+    public EducationDto addEducation(Long userId, EducationDto educationDto) throws DataValidationException {
+        checkYearFrom(educationDto);
+        checkUserIdNull(userId);
```
этот метод есть в `saveEducation()` и поэтому его лучше отсюда убрать и оставить только в одном месте, в методе `saveEducation()`

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public EducationDto updateEducation(Long userId, EducationDto educationDto) throws DataValidationException {
+        checkYearFrom(educationDto);
```
этот метод есть в `saveEducation()` и поэтому его лучше отсюда убрать и оставить только в одном месте, в методе `saveEducation()`

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,80 @@
+
+    public EducationDto updateEducation(Long userId, EducationDto educationDto) throws DataValidationException {
+        checkYearFrom(educationDto);
+        checkUserIdNull(userId);
```
этот метод есть в `saveEducation()` и поэтому его лучше отсюда убрать и оставить только в одном месте, в методе `saveEducation()`

re1971reg commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,80 @@
+        checkYearFrom(educationDto);
+        checkUserIdNull(userId);
+
+        if (userId != educationDto.getId()) {
```
может возникнуть ситуация, когда в переменной userId будет null, поэтому вначале делается проверка, что userId != null, а потом уже можно проверить userId != educationDto.getId()

FilippovichSergey committed:
class EducationService, methods checkUserIdNull, addEducation, updateEducation, saveEducation, checkUserIdEmpty. Updated.

FilippovichSergey committed:
class EducationService. Updated spaces.

FilippovichSergey committed:
class EducationService. Updated style.

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/controller/education/EducationController.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final EducationService educationService;
+
+    public EducationDto addEducation(Long userId, EducationDto educationDto) throws DataValidationException {
```
throws DataValidationException не должно быть у контроллеров. Это следует удалить

FilippovichSergey replied:
Почему throws DataValidationException не должно быть у контроллеров?
Надо заменить на try/catch?

SnurnicynKonstantin replied:
Надо в DataValidationException заменить Throwable на RuntimeException. 
Таким образом мы говорим, что ошибка может возникнуть в результате работы, но не обязательно возникнет.
Если оставим Throwable, то их придется покрывать try/catch. Spring если ловит RuntimeException ошибки, пользователю при запросе возвращает 500. 
А если при Throwable в try/catch оборачивать, то что пользователю возвращать? Опять же ошибку и код придумывать для нее? На самом деле мы так и сделали. В мастере уже есть обработчик вызыванных исключений - https://github.com/CorporationX/user_service/blob/chimera-master-stream10/src/main/java/school/faang/user_service/rest/ExceptionHandlerControllerAdvice.java
Он работает как try/catch. Сам ловит исключения, которые ты вызываешь в коде и возвращает нужный код ответа пользователю. Но на 5м спринте мы это не проходили. Так что я его не требую. А вот исключения в REST сервисе надо делать RuntimeException. По крайней мере связанные с подобного рода валидациями.

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/controller/education/EducationController.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final EducationService educationService;
+
+    public EducationDto addEducation(Long userId, EducationDto educationDto) throws DataValidationException {
```
Сейчас используешь EducationDto для запросов и ответов. Но для запроса не нужно поле id. 
Советую оздат отдельное дто для этого метода - EducationAddRequestDto.
А EducationDto переименовать в EducationResponseDto. И ее использовать для ответов

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/dto/EducationDto.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.Data;
+
+@Data
+public class EducationDto {
```
Можно в record переделать?

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,72 @@
+    private final EducationRepository educationRepository;
+    private final EducationMapper educationMapper;
+
+    public void checkYearFrom(EducationDto educationDto) throws DataValidationException {
```
throws DataValidationException то же следует удалить

FilippovichSergey replied:
Почему?

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    public User checkUserIdEmpty(Long userId) throws DataValidationException {
+        return userRepository.findById(userId)
```
Старайся избегать использования чужих репозиториев. Если этот сервис про Education, то по хорошему он должен только educationRepository использовать. А что бы юзера получить используй UserService, там уже есть метод getUserById()

FilippovichSergey committed:
1) removed throws DataValidationException
2) split DTO to two DTOs
3) changed DTO to record

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/controller/education/EducationController.java:
```diff
@@ -12,15 +12,23 @@ public class EducationController {
+    public EducationResponseDto addEducation(Long userId, EducationResponseDto educationResponseDto) {
+        try {
+            return educationService.addEducation(userId, educationResponseDto);
+        } catch (DataValidationException e) {
```
catch можешь удалить. У нас появился класс, который отлавливает DataValidationException сам.
https://github.com/CorporationX/user_service/blob/chimera-master-stream10/src/main/java/school/faang/user_service/rest/ExceptionHandlerControllerAdvice.java 
На 5м спринте пока это не проходили. Так что просто удали кэтч

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/education/EducationService.java:
```diff
@@ -11,60 +11,62 @@
-        throw new DataValidationException("Your year is greater than the current year.");
+        try {
+            throw new DataValidationException("Your year is greater than the current year.");
+        } catch (DataValidationException e) {
```
У тебя DataValidationException и так RuntimeException. Оставь throw new DataValidationException("Your year is greater than the current year."); без try/catch

SnurnicynKonstantin reviewed (commented):
Сделай rebase на мастер пожалуйста. У тебя с ним конфликты, которые надо решить
